datafusion/functions/src/unicode/lpad.rs

```diff
@@ -76,300 +87,450 @@ impl ScalarUDFImpl for LPadFunc {
+                args[0].as_string_view(),
+                length_array,
+                &args[2],
+            ),
+            _ => unreachable!(),
+        },
+        other => {
+            exec_err!("Unsupported data type {other:?} for function lpad")
+        }
+    }
+}
 
-                        let length = if length < 0 { 0 } else { length as usize };
-                        if length == 0 {
-                            Ok(Some("".to_string()))
+fn lpad_with_replace<'a, V, T: OffsetSizeTrait>(
+    string_array: V,
+    length_array: &Int64Array,
+    fill_array: &'a ArrayRef,
+) -> Result<ArrayRef>
+where
+    V: StringArrayType<'a>,
+{
+    match fill_array.data_type() {
+        Utf8 => lpad_impl::<V, &GenericStringArray<i32>, T>(
+            string_array,
+            length_array,
+            Some(fill_array.as_string::<i32>()),
+        ),
+        LargeUtf8 => lpad_impl::<V, &GenericStringArray<i64>, T>(
+            string_array,
+            length_array,
+            Some(fill_array.as_string::<i64>()),
+        ),
+        Utf8View => lpad_impl::<V, &StringViewArray, T>(
+            string_array,
+            length_array,
+            Some(fill_array.as_string_view()),
+        ),
+        other => {
+            exec_err!("Unsupported data type {other:?} for function lpad")
+        }
+    }
+}
+
+fn lpad_impl<'a, V, V2, T>(
+    string_array: V,
+    length_array: &Int64Array,
+    fill_array: Option<V2>,
+) -> Result<ArrayRef>
+where
+    V: StringArrayType<'a>,
+    V2: StringArrayType<'a>,
+    T: OffsetSizeTrait,
+{
+    if fill_array.is_none() {
+        let result = string_array
+            .iter()
+            .zip(length_array.iter())
+            .map(|(string, length)| match (string, length) {
+                (Some(string), Some(length)) => {
+                    if length > i32::MAX as i64 {
+                        return exec_err!("lpad requested length {length} too large");
+                    }
+
+                    let length = if length < 0 { 0 } else { length as usize };
+                    if length == 0 {
+                        Ok(Some("".to_string()))
+                    } else {
+                        let graphemes = string.graphemes(true).collect::<Vec<&str>>();
+                        if length < graphemes.len() {
+                            Ok(Some(graphemes[..length].concat()))
                         } else {
-                            let graphemes = string.graphemes(true).collect::<Vec<&str>>();
-                            if length < graphemes.len() {
-                                Ok(Some(graphemes[..length].concat()))
-                            } else {
-                                let mut s: String = " ".repeat(length - graphemes.len());
-                                s.push_str(string);
-                                Ok(Some(s))
-                            }
+                            let mut s: String = " ".repeat(length - graphemes.len());
```

Comment:
This PR just follows the existing pattern so I don't think any changes are needed, but I think using a `GenericStringBuilder` would be much faster here (as it wouldn't allocate a string for each output array element)

Reply:
I did a test locally with using `GenericStringBuilder` vs what is in this PR and the differences are within +/- 10% for 1024 and 2048 element arrays for each of the 3 string types. Worth noting though is that timing between runs on my laptop is not really stable at all (+/- 10% or so between runs of the exact same code).

Interestingly the utf8view seems to be consistently slightly slower than just plain utf8 at least in this function.
```
lpad function/utf8 type/2048
                        time:   [715.78 µs 718.43 µs 721.89 µs]
                        
lpad function/largeutf8 type/2048
                        time:   [745.44 µs 759.62 µs 779.84 µs]
                       
lpad function/stringview type/2048
                        time:   [734.25 µs 751.13 µs 773.67 µs]
                       
  ```